woocommerce/woocommerce-ios
# [Mobile Payments] Fix Tap to Pay reconnection on foreground, after previous try fails

Closes: #11050 


## Description


Previously, if an automatic reconnection to the Tap to Pay reader failed when the app was brought in to the foreground, the app would never try to reconnect again on subsequent foreground attempts, until it was either force quit or kicked from memory.

This resulted in merchants having to wait longer to be able to take a payment when they opened the app.

This commit changes the connection controller to allow reuse, by adopting the completion handler of a subsequent call to `searchAndConnect` without resetting its connection attempt, which would usually fail. When a connection attempt fails, it resets to the idle state, to allow it to be reused.

Additionally, the reconnection controller now reuses its connection controller when it still exists from a previous automatic reconnection attempt – which only happens the previous attempt failed.


## Testing instructions


Using a WooPayments store in the US or UK

1. Launch the app from Xcode
2. Take a Tap to Pay payment – this first transaction on the device makes the app eligible for reconnection.
3. Background the app
4. Foreground the app and immediately background it – check the console output, you should see a Tracks event for `card_reader_connection_failed` with "There was an issue preparing to use Tap to Pay on iPhone – please try again.". Repeat if you see only success events; you need to be very quick.
5. Navigate to `Menu > Payments > Try Out Tap to Pay on iPhone`
6. Observe that you see the Set Up prompt on this screen, not the green checkmark with a `Continue` button indicating that you're already connected to the reader.
7. Background and foreground the app again; normal speed this time
8. Go back to the `Try Out Tap to Pay on iPhone` screen: observe that you've been automatically reconnected. (Do give it 5-10s before you go here, reconnection does take some time)
9. Observe that the console contains a `card_reader_connection_success` event after the previous failure.

Note that if if failed for another reason on the second launch, you won't be connected, and you'll see `card_reader_connection_failed` again in the console. This still represents a fix; it tried to reconnect, where previously it wouldn't have.

## Screenshots


### Before

https://github.com/woocommerce/woocommerce-ios/assets/2472348/9a7ba96e-7072-4357-893a-24ba6195c192


### After

https://github.com/woocommerce/woocommerce-ios/assets/2472348/e3487a3a-fe40-48e9-a7a8-3d6631211abb


---
